### PR TITLE
[TS migration] Migrate 'ReportActionItemParentAction.js' component to TypeScript

### PR DESCRIPTION
**Pull request by sourcecodedeveloper**
### Fixed Issues

$ https://github.com/Expensify/App/issues/31962
PROPOSAL: https://github.com/Expensify/App/issues/31962#issuecomment-1883524928

### Tests

1. Create chat with another user
2. Click Reply in thread and write message.
3. The ReportActionItemParentAction component is loaded when replying in thread messages
4. Verify the app does not crash after reply in thread messages.

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android.webm](https://github.com/Expensify/App/assets/141340971/c09d7962-a0ac-4322-a7aa-804c2b8628e1)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android.web.webm](https://github.com/Expensify/App/assets/141340971/d2524da3-9b2d-427d-8de6-b9c64d1ad063)

</details>

<details>
<summary>iOS: Native</summary>

[ios.webm](https://github.com/Expensify/App/assets/141340971/7f4d6784-bfa4-47c8-9f09-577bf1ff6239)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

[ios.web.webm](https://github.com/Expensify/App/assets/141340971/91867ef2-e006-43eb-a143-f22c4909d241)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

[web.webm](https://github.com/Expensify/App/assets/141340971/bc19ed6e-12e9-4d7d-b1af-d9d5af112a7f)

</details>

<details>
<summary>MacOS: Desktop</summary>

[desktop.webm](https://github.com/Expensify/App/assets/141340971/0f8dc2a3-18f5-459b-848b-9292a98a751b)

</details>

### CODE REVIEW

**Comment by allroundexperts** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by allroundexperts** [comment]
@sourcecodedeveloper Please complete your checklist. All the screenshots are missing.

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+function ReportActionItemParentAction({report, parentReportActions, isSmallScreenWidth, shouldHideThreadDividerLine, shouldDisplayNewMarker}: ReportActionItemParentActionProps) {
+    const styles = useThemeStyles();
+    const StyleUtils = useStyleUtils();
+    if (!parentReportActions) {
+        return;
+    }
+    if (!report) {
+        return;
+    }
```
Comment (allroundexperts) [inline src/pages/home/report/ReportActionItemParentAction.tsx:52]: Why are these added? And why have you removed the defaults that were defined in the previous implementation?

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    const parentReportAction = parentReportActions[`${report?.parentReportActionID}`];
+    const shouldDisableOpacity = parentReportAction?.pendingAction ?? false;
+    const pendingAction = report?.pendingFields?.addWorkspaceRoom ?? report?.pendingFields?.createChat;
+    const errors = report?.errorFields?.addWorkspaceRoom ?? report?.errorFields?.createChat;
```
Comment (allroundexperts) [inline src/pages/home/report/ReportActionItemParentAction.tsx:56]: I don't see any reason for saving these into constants. I think we can use them directly as they were previously.

**Review by allroundexperts — CHANGES_REQUESTED** [review]
Requires changes

File: src/pages/home/report/ReportActionItemParentAction.tsx (hunk 1 above)
Comment (sourcecodedeveloper) [reply src/pages/home/report/ReportActionItemParentAction.tsx]: I removed the if conditions. The TS_STYLE.md suggests not to use defaultProps and also other tsx files do not have defaultprops

File: src/pages/home/report/ReportActionItemParentAction.tsx (hunk 2 above)
Comment (sourcecodedeveloper) [reply src/pages/home/report/ReportActionItemParentAction.tsx]: I removed the constants.

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+        /** The id of the report */
+        reportID: string;
+
+        /** The id of the parent report */
+        // eslint-disable-next-line react/no-unused-prop-types
+        parentReportID: string;
+    };
+
+function ReportActionItemParentAction({report, parentReportActions, isSmallScreenWidth, shouldHideThreadDividerLine, shouldDisplayNewMarker}: ReportActionItemParentActionProps) {
```
Comment (allroundexperts) [inline src/pages/home/report/ReportActionItemParentAction.tsx:44]: @sourcecodedeveloper Can you please add the default props that were there before here?

Comment (allroundexperts) [reply src/pages/home/report/ReportActionItemParentAction.tsx:44]: As an example, the default prop for report should be defined like:
```
function ReportActionItemParentAction({report = {}, parentReportActions, isSmallScreenWidth, shouldHideThreadDividerLine, shouldDisplayNewMarker}: ReportActionItemParentActionProps) {
```

Comment (sourcecodedeveloper) [reply src/pages/home/report/ReportActionItemParentAction.tsx:44]: report = {} gives error,  Property 'reportID' is missing in type '{}' but required in type 'Report'.

I added default props for
    parentReportActions = {},
    shouldHideThreadDividerLine = false,

Comment (allroundexperts) [reply src/pages/home/report/ReportActionItemParentAction.tsx:44]: But why does it give that error? In the previous implementation, we were using the same `{}` as the default prop.

**Comment by allroundexperts** [comment]
Conflicts!

File: src/pages/home/report/ReportActionItemParentAction.tsx (hunk 3 above)
Comment (sourcecodedeveloper) [reply src/pages/home/report/ReportActionItemParentAction.tsx]: I resolved the error, reportID also needed to be set with default value

**Comment by sourcecodedeveloper** [comment]
@allroundexperts I added screenshots.

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        // eslint-disable-next-line react/no-unused-prop-types
+        reportID: string;
+
+        /** The id of the parent report */
+        // eslint-disable-next-line react/no-unused-prop-types
+        parentReportID: string;
+    };
+
+function ReportActionItemParentAction({
+    report = {reportID: ''},
```
Comment (allroundexperts) [inline src/pages/home/report/ReportActionItemParentAction.tsx:49]: @sourcecodedeveloper Why was `reportID` not included in the previous default props?

Comment (sourcecodedeveloper) [reply src/pages/home/report/ReportActionItemParentAction.tsx:49]: @allroundexperts The previous default props props.report had type any, so the required reportID was not checked for, the tsx uses ReportActionItemParentActionProps instead of any.

**Comment by sourcecodedeveloper** [comment]
@allroundexperts The author checklist is complete and waiting on Reviewer checklist.

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import useThemeStyles from '@hooks/useThemeStyles';
+import * as ReportActionsUtils from '@libs/ReportActionsUtils';
+import * as Report from '@userActions/Report';
+import ONYXKEYS from '@src/ONYXKEYS';
+import type * as OnyxTypes from '@src/types/onyx';
+import AnimatedEmptyStateBackground from './AnimatedEmptyStateBackground';
+import ReportActionItem from './ReportActionItem';
+
+type ReportActionItemParentActionOnyxProps = {
+    /** ONYX PROPS */
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemParentAction.tsx:18]:
```suggestion
```

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    /** The report currently being looked at */
+    report: OnyxEntry<OnyxTypes.Report>;
+
+    /** The actions from the parent report */
+    // TO DO: Replace with HOC https://github.com/Expensify/App/issues/18769.
+    parentReportActions: OnyxEntry<OnyxTypes.ReportActions>;
+};
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemParentAction.tsx:26]:
```suggestion
};

```

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+type ReportActionItemParentActionProps = WithLocalizeProps &
+    WindowDimensionsProps &
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemParentAction.tsx:28]: Remove WithLocalizeProps and WindowDimensionsProps, use hooks instead.

```suggestion
type ReportActionItemParentActionProps = 
```

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    ReportActionItemParentActionOnyxProps & {
+        /** Flag to show, hide the thread divider line */
+        shouldHideThreadDividerLine: boolean;
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemParentAction.tsx:31]: This prop should be optional

File: src/pages/home/report/ReportActionItemParentAction.tsx (hunk 4 above)
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemParentAction.tsx:49]: You should be able to remove this default value completely:
```suggestion
    report,
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by blazejkustra** [comment]
@sourcecodedeveloper Goo work overall! Please rename the PR to [TS migration] Migrate 'ReportActionItemParentAction.js' component to TypeScript

**Comment by sourcecodedeveloper** [comment]
@blazejkustra Thanks, I renamed PR title.

**Comment by sourcecodedeveloper** [comment]
@blazejkustra PR is ready for review.

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+export default withOnyx<ReportActionItemParentActionProps, ReportActionItemParentActionOnyxProps>({
+    report: {
+        key: ({reportID}) => `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+    },
+    parentReportActions: {
+        key: ({parentReportID}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${parentReportID}`,
+        canEvict: false,
+    },
+})(ReportActionItemParentAction);
+export type {ReportActionItemParentActionProps, ReportActionItemParentActionOnyxProps};
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemParentAction.tsx:92]: If props aren't used outside of this file, you don't need to export them

**Review by blazejkustra — APPROVED** [review]
Once minor comment, LGTM

**Comment by sourcecodedeveloper** [comment]
@allroundexperts PR is ready for review.

**Review by allroundexperts — APPROVED** [review]
Looks good!

**Review by yuwenmemon — APPROVED** [review]
